packages/next/src/build/webpack-config.ts

```diff
@@ -1400,41 +1401,72 @@ export default async function getBaseWebpackConfig(
     // Also disable esm request when appDir is enabled
     const isEsmRequested = dependencyType === 'esm'
 
+    /**
+     * @param {string} localRes the full path to the file
+     * @returns {string | undefined} the externalized path
+     * @description returns an externalized path if the file is a Next.js file and ends with either `.shared-runtime.js` or `.external.js`
+     * This is used to ensure that files used across the rendering runtime(s) and the user code are one and the same. The logic in this function
+     * will rewrite the require to the correct bundle location depending on the layer at which the file is being used.
+     */
     const isLocalCallback = (localRes: string) => {
-      // Makes sure dist/shared and dist/server are not bundled
-      // we need to process shared `router/router`, `head` and `dynamic`,
-      // so that the DefinePlugin can inject process.env values.
+      const pathSeparators = '[/\\\\]'
+      const optionalEsmPart = `((${pathSeparators}esm)?${pathSeparators})`
+      const sharedRuntimeFileEnd = '(\\.shared-runtime(\\.js)?)$'
+      const externalFileEnd = '(\\.external(\\.js)?)$'
+      const nextDist = `next${pathSeparators}dist`
+
+      const sharedRuntimePattern = new RegExp(
+        `${nextDist}${optionalEsmPart}.*${sharedRuntimeFileEnd}`
+      )
+      const externalPattern = new RegExp(
```

Comment:
Since these regexes seem to be mostly hardcoded could you move them to the module scope? This function is called very often during compilation so it'll help avoid some code compilation.